trieTools/tokensExporter/main.go

```diff
@@ -0,0 +1,251 @@
+	if err != nil {
+		return err
+	}
+
+	err = accDb.RecreateTrie(mainRootHash)
+	if err != nil {
+		return err
+	}
+
+	numAccountsOnMainTrie := 0
+	addressTokensMap := make(map[string]map[string]struct{})
+	for keyValue := range ch {
+		address, found := getAddress(keyValue)
+		if !found {
+			continue
+		}
+
+		numAccountsOnMainTrie++
+
+		account, errGetAccount := accDb.GetExistingAccount(address)
+		if errGetAccount != nil {
+			return errGetAccount
+		}
+
+		esdtTokens, errGetESDT := getAllESDTTokens(account)
+		if errGetESDT != nil {
+			return errGetESDT
+		}
+
+		if len(esdtTokens) > 0 {
+			encodedAddress := addressConverter.Encode(address)
+			addressTokensMap[encodedAddress] = esdtTokens
+		}
+	}
+
+	encodedSysAccAddress := addressConverter.Encode(vmcommon.SystemAccountAddress)
+	log.Info("parsed main trie",
+		"num accounts", numAccountsOnMainTrie,
+		"num accounts with tokens", len(addressTokensMap),
+		"num tokens in system account address", len(addressTokensMap[encodedSysAccAddress]))
+
+	_, found := addressTokensMap[encodedSysAccAddress]
+	if !found {
+		log.Warn(fmt.Sprintf("system account address(%s) not found, input dbs might be incomplete/corrupted", encodedSysAccAddress))
+	}
+
+	return saveResult(addressTokensMap, flags.Outfile)
+}
+
+func getAddress(kv core.KeyValueHolder) ([]byte, bool) {
+	userAccount := &state.UserAccountData{}
+	errUnmarshal := trieToolsCommon.Marshaller.Unmarshal(userAccount, kv.Value())
+	if errUnmarshal != nil {
+		// probably a code node
+		return nil, false
+	}
+	if len(userAccount.RootHash) == 0 {
+		return nil, false
+	}
+
+	return kv.Key(), true
+}
+
+func saveResult(addressTokensMap map[string]map[string]struct{}, outfile string) error {
+	jsonBytes, err := json.MarshalIndent(addressTokensMap, "", " ")
+	if err != nil {
+		return err
+	}
+
+	log.Info("writing result in", "file", outfile)
+	err = ioutil.WriteFile(outfile, jsonBytes, fs.FileMode(outputFilePerms))
+	if err != nil {
+		return err
+	}
+
+	log.Info("finished exporting address-tokens map")
+	return nil
+}
+
+func getAllESDTTokens(account vmcommon.AccountHandler) (map[string]struct{}, error) {
+	userAccount, ok := account.(state.UserAccountHandler)
+	if !ok {
+		return nil, fmt.Errorf("could not convert account to user account, address = %s",
+			hex.EncodeToString(account.AddressBytes()))
+	}
+
+	allESDTs := make(map[string]struct{})
+	if check.IfNil(userAccount.DataTrie()) {
+		return allESDTs, nil
+	}
+
+	rootHash, err := userAccount.DataTrie().RootHash()
+	if err != nil {
+		return nil, err
+	}
+
+	chLeaves := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = userAccount.DataTrie().GetAllLeavesOnChannel(chLeaves, context.Background(), rootHash)
+	if err != nil {
+		return nil, err
+	}
+
+	esdtPrefix := []byte(core.ElrondProtectedKeyPrefix + core.ESDTKeyIdentifier)
+	for leaf := range chLeaves {
+		if !bytes.HasPrefix(leaf.Key(), esdtPrefix) {
+			continue
+		}
+
+		tokenKey := leaf.Key()
```

Comment:
you might add an additional step: try to unmarshal it. in the future, when the new metadata storage will be live, we'll have to filter the results between old storage model vs new storage model. at least add a todo

Reply:
Added a todo